### ISSUE TITLE
Distributed Training on multiple nodes - process gets stuck after initializing grpc channel

### ISSUE THREAD

**Issue by hroetsc** [tensorflow/tensorflow#39336]
**System information**
- TF v.2.1
- Linux-based HPC
- Snakemake workflow manager
- Slurm as scheduler
- TensorFlow installed from (source or binary): virtual conda environment
- Python version: 3.7.6
- CUDA/cuDNN version: 10.1
- GPU model and memory: GeForce GTX 980 computeCapability: 5.2,  coreClock: 1.2405GHz coreCount: 16 deviceMemorySize: 3.95GiB deviceMemoryBandwidth: 208.91GiB/s

**Describe the current behavior**
Hi there,  
My aim is to train a neural net on multiple nodes and GPUs on a HPC. Therefore, I am using TF's `MultiWorkerMirroredStrategy` and the `SlurmClusterResolver` to get the configuration of my nodes and to set the TF_CONFIG variable.  
  
However, when trying to connect to the cluster using:  
```
tf.config.experimental_connect_to_cluster(resolver,
                                            job_name = 'worker',
                                            task_index = cfg['task']['index'],
                                            protocol = 'grpc')
```  
the process gets stuck (please see the log below).  
When I `ssh` into the nodes I see my processes there but they are all sleeping. GPUs aren't used either.  
  
Has anyone experience with this kind of setup and can provide help?  
  
If I don't use the `experimental_connect_to_cluster` every node is executing the job independently, instead of working together.  
I know that TF_CONFIG should be set before calling the `MultiWorkerMirroredStrategy` but this caused a `RunTimeError`.  
I also played around with the position of GPU initialization in the code - this doesn't seem to have an effect.  
  
Note that I am using the most recent version of the `SlurmClusterResolver`, I basically copied the script from the GitHub and call it using `slurm_cluster_resolver.SlurmClusterResolver`

**Standalone code to reproduce the issue**
```
import tensorflow as tf

tf.keras.backend.clear_session()
# print all tensor allocations
tf.debugging.set_log_device_placement(True)
tf.random.set_seed(42)

# instantiate strategy at program startup to prevent RuntimeError
print("-------------------------------------------------------------------------")

print("DEFINE DISTRIBUTED TRAINING STRATEGY")
# only RING communication uses grpc protocols

multiworker_strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy(communication=tf.distribute.experimental.CollectiveCommunication.NCCL)


import os
import sys
import json
import gc
import numpy as np
import pandas as pd



print("-------------------------------------------------------------------------")

print('CLUSTER CONFIGURATION')

def set_tf_config(resolver, environment=None):
    """Set the TF_CONFIG env variable from the given cluster resolver"""
    cfg = {
        'cluster': resolver.cluster_spec().as_dict(),
        'task': {
            'type': resolver.get_task_info()[0],
            'index': resolver.get_task_info()[1],
        },
        'rpc_layer': resolver.rpc_layer,
    }
    if environment:
        cfg['environment'] = environment
    os.environ['TF_CONFIG'] = json.dumps(cfg)

    return cfg


# there must be one GPU for every task
resolver = slurm_cluster_resolver.SlurmClusterResolver(port_base = 11214, gpus_per_task=1, tasks_per_node=2)

print("-------------------------------------------------------------------------")

cfg = set_tf_config(resolver)
tf.print(cfg)
print(cfg)


print('CONNECT TO CLUSTER')

tf.config.experimental_connect_to_cluster(resolver,
                                            job_name = 'worker',
                                            task_index = cfg['task']['index'],
                                            protocol = 'grpc')

print("-------------------------------------------------------------------------")

# doesn't matter if at the beginning or not
print("GPU CONFIGURATION")
# allow memory growth

gpus = tf.config.experimental.list_physical_devices('GPU')

print('INITIALIZE GPUs')
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)

if gpus:
    tf.config.experimental.set_visible_devices(gpus, 'GPU')
    print(gpus)

print("-------------------------------------------------------------------------")



```
**Other info / logs**
```
2020-05-09 08:40:42.220277: I tensorflow/core/common_runtime/eager/execute.cc:573] Executing op StringFormat in device /job:localhost/replica:0/task:0/device:CPU:0
2020-05-09 08:40:42.220277: I tensorflow/core/common_runtime/eager/execute.cc:573] Executing op StringFormat in device /job:localhost/replica:0/task:0/device:CPU:0
2020-05-09 08:40:42.220607: I tensorflow/core/common_runtime/eager/execute.cc:573] Executing op PrintV2 in device /job:localhost/replica:0/task:0/device:CPU:0
2020-05-09 08:40:42.220639: I tensorflow/core/common_runtime/eager/execute.cc:573] Executing op PrintV2 in device /job:localhost/replica:0/task:0/device:CPU:0
{'cluster': {'worker': ['dge10:11214',
                'dge10:11215',
                'dge12:11214',
                'dge12:11215',
                'dge13:11214',
                'dge13:11215',
                'dge14:11214',
                'dge14:11215',
                'dge15:11214',
                'dge15:11215',
                'dge9:11214',
                'dge9:11215']},
'rpc_layer': 'grpc',
'task': {'index': 5, 'type': 'worker'}}
{'cluster': {'worker': ['dge10:11214',
                'dge10:11215',
                'dge12:11214',
                'dge12:11215',
                'dge13:11214',
                'dge13:11215',
                'dge14:11214',
                'dge14:11215',
                'dge15:11214',
                'dge15:11215',
                'dge9:11214',
                'dge9:11215']},
'rpc_layer': 'grpc',
'task': {'index': 4, 'type': 'worker'}}
2020-05-09 08:40:42.223059: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1096] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-05-09 08:40:42.223095: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1096] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-05-09 08:40:42.223113: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102]
2020-05-09 08:40:42.223157: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102]
2020-05-09 08:40:42.226758: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:300] Initialize GrpcChannelCache for job worker -> {0 -> dge10:11214, 1 -> dge10:11215, 2 -> dge12:11214, 3 -> dge12:11215, 4 -> localhost:11214, 5 -> dge13:11215, 6 -> dge14:11214, 7 -> dge14:11215, 8 -> dge15:11214, 9 -> dge15:11215, 10 -> dge9:11214, 11 -> dge9:11215}
2020-05-09 08:40:42.226828: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:300] Initialize GrpcChannelCache for job worker -> {0 -> dge10:11214, 1 -> dge10:11215, 2 -> dge12:11214, 3 -> dge12:11215, 4 -> dge13:11214, 5 -> localhost:11215, 6 -> dge14:11214, 7 -> dge14:11215, 8 -> dge15:11214, 9 -> dge15:11215, 10 -> dge9:11214, 11 -> dge9:11215}
```

**Comment by Saduf2019** [comment]
@hroetsc 
I ran the code shared above but face a different error, please refer to this [gist here](https://colab.sandbox.google.com/gist/Saduf2019/2ec7fc4324cb2118d6aa2c4be1f92e3d/untitled176.ipynb) for the same.

**Comment by hroetsc** [comment]
instead of `slurm_cluster_resolver.SlurmClusterResolver()` you could just run `tf.distribute.cluster_resolver.SlurmClusterResolver()` or take the most recent version from the GitHub, copy it into you directory and name it slurm_cluster_resolver.py

**Comment by yuefengz** [comment]
`experimental_connect_to_cluster ` is not necessary for `MultiWorkerMirroredStrategy`. What are the issues to set TF_CONFIG?

**Comment by hroetsc** [comment]
manually set TF_CONFIG variables were not maintained when i logged out of the node... Since the cluster is very large it is also very inconvenient to set them manually

**Comment by yuefengz** [comment]
`MultiWorkerMirroredStrategy` also accepts `ClusterResolver` as input. You can pass your resolver into it instead of using experimental_connect_to_cluster. Note the `ClusterResolver` passed in must have `task_type` and `task_index` set.

**Comment by hroetsc** [comment]
i ended up getting a `RunTimeError` when calling the `MultiWorkerMirroredStrategy` after the `Cluster Resolver`

**Comment by yuefengz** [comment]
Did you do something with your cluster_resolver? Maybe you triggered something that initialized some global state. Note you should never call `experimental_connect_to_cluster` for `MultiWorkerMirroredStrategy`.

Could you show us the code how you construct your cluster_resolver and how you pass it to `MultiWorkerMirroredStrategy`?

**Comment by hroetsc** [comment]
i solved the problem - using Horovod with Tensorflow instead of Tensorflow's dirstributed training. But thank you for your help!
